Update tutorials

tkoskela opened the pull request:
I've added the ReFrame tutorial and fixed the most obvious formatting issues from hackmd. I've also added separate tutorials for getting started, postprocessing and profiling, based on the materials for CIUK.

There are a few rough edges and some of the Cosma output tabs are missing, but I think I'm mostly happy with the content.

You can preview the website by clicking on the [details of the publish action](https://ukri-excalibur.github.io/excalibur-tests/preview/PR344/)

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+- ReFrame uses decorators to mark classes as tests. 
+- This marks `class HelloTest` as a `rfm.simple_test`.
+- ReFrame tests ultimately derive from `RegressionTest`. There are others derived classes such as `RunOnlyRegressionTest`, we get to those later.
```
```suggestion
- ReFrame tests ultimately derive from `RegressionTest`. There are other derived classes such as `RunOnlyRegressionTest`, we get to those later.
```

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+----
+
+## Builting programming environment
```
```suggestion
## Builtin programming environment
```

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+---
+
+## Configuring ReFrame for HPC systems (Cosma)
+> In ReFrame, all the details of the various interactions of a test with the system environment are handled transparently and are set up in its configuration file.
```
```suggestion
In ReFrame, all the details of the various interactions of a test with the system environment are handled transparently and are set up in its configuration file.
```

tkoskela replied:
The quotation `>` is for stuff that is directly copied from the ReFrame tutorial

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+## Configuring ReFrame for HPC systems (Cosma)
+> In ReFrame, all the details of the various interactions of a test with the system environment are handled transparently and are set up in its configuration file.
+- [Configuration](https://reframe-hpc.readthedocs.io/en/v4.5.2/tutorial_basics.html#more-of-hello-world)
```
Should we point to [this link](https://reframe-hpc.readthedocs.io/en/v4.5.2/configure.html) instead?

tkoskela replied:
I think I wanted to point to the block of code after 

> Here is what we need to type:

but there is no direct link to it. The one I have is the header of the section it's in

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+- `reframe --show-config`
+- Builtin programming environment uses `cc` to compile
+
+---
```
When are you going to show how to run this benchmark?

tkoskela replied:
Now!

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+Performance tests capture data in performance variables. For simplicity, we use the [STREAM benchmark](https://github.com/jeffhammond/STREAM) as an example. It is the de facto memory bandwidth benchmark.
+
+To record the performance of the benchmark, ReFrame should extract a figure of merit from the output of the test. A function decorated with the `@performance_function` decorator extracts or computes a performance metric from the test’s output.
```
Remove this paragraph, it's duplicated in the "Add Performance Pattern Check" section below

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+To record the performance of the benchmark, ReFrame should extract a figure of merit from the output of the test. A function decorated with the `@performance_function` decorator extracts or computes a performance metric from the test’s output.
+
+> In this example, we extract four performance variables, namely the memory bandwidth values for each of the “Copy”, “Scale”, “Add” and “Triad” sub-benchmarks of STREAM, where each of the performance functions use the [`extractsingle()`](https://reframe-hpc.readthedocs.io/en/latest/deferrable_functions_reference.html#reframe.utility.sanity.extractsingle) utility function. For each of the sub-benc ...
```
```suggestion
In this example, we extract four performance variables, namely the memory bandwidth values for each of the “Copy”, “Scale”, “Add” and “Triad” sub-benchmarks of STREAM, where each of the performance functions use the [`extractsingle()`](https://reframe-hpc.readthedocs.io/en/latest/deferrable_functions_reference.html#reframe.utility.sanity.extractsingle) utility function. For each of the sub-benchmarks we extract the “Best Rate MB/s” column of the output (see below) and we convert that to a float.
```

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+----
+
+## Performance Pattern Check
```
This doesn't need to be a separate section, it's part of the above.

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+### Perflogs
+
+- Perflogs are output in `perflogs/<system>/<partition>`
```
What is a perflog?

tkoskela replied:
It's a secret! I added some hints

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+- Perflogs are output in `perflogs/<system>/<partition>`
+- By default a lot of information is printed. This can be customized in the configuration file. More on this later.
+- By default not much information about build step, has to be linked back to build environment
```
I don't understand this sentence.

tkoskela replied:
Does it make more sense now?

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
````diff
@@ -0,0 +1,409 @@
+}
+```
+
+> The performance reference tuple consists of the reference value, the lower and upper thresholds expressed as fractional numbers relative to the reference value, and the unit of measurement. If any of the thresholds is not relevant, None may be used instead. Also, the units in this reference variable are entirely optional, since they were already provided through the @performance_function decora ...
````
```suggestion
The performance reference tuple consists of the reference value, the lower and upper thresholds expressed as fractional numbers relative to the reference value, and the unit of measurement. If any of the thresholds is not relevant, `None` may be used instead. Also, the units in this reference variable are entirely optional, since they were already provided through the @performance_function decorator.
```

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+----
+
+### Parametrized tests
```
```suggestion
## Parametrized tests
```

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
````diff
@@ -0,0 +1,409 @@
+pip install reframe-hpc==4.5.2
+```
+
+Alternatively, you can
````
Say something here like cloning the ReFrame repo if you want access to the tutorials - paths are referenced below.

tkoskela replied:
You also get the tutorials if you pip install. But yeah maybe worth saying the tutorials are in there.

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+----
+
+### [Make](https://reframe-hpc.readthedocs.io/en/v4.5.2/tutorial_advanced.html#more-on-building-tests) 
```
Do we need to duplicate the link here?

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+
+- Tutorial in `tutorials/advanced/makefiles/maketest.py`. 
+
+> you must set the executable attribute of the test, because ReFrame cannot know what is the actual executable to be run. We then set the build system to Make and set the preprocessor flags as we would do with the SingleSource build system.
```
This paragraph makes no sense on its own. Either remove or explain.
Also, it's formatted as a comment.

tkoskela replied:
I included the whole paragraph in the quote, does it make more sense?

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+----
+
+### [Autotools](https://reframe-hpc.readthedocs.io/en/v4.5.2/tutorial_advanced.html#adding-a-configuration-step-before-compiling-the-code)
+> It is often the case that a configuration step is needed before compiling a code with make. To address this kind of projects, ReFrame aims to offer specific abstractions for “configure-make” style of build systems. It supports CMake-based projects through the CMake build system, as well as Autotools-based projects through the Autotools build system.
```
```suggestion
It is often the case that a configuration step is needed before compiling a code with make. To address this kind of projects, ReFrame aims to offer specific abstractions for “configure-make” style of build systems. It supports CMake-based projects through the CMake build system, as well as Autotools-based projects through the Autotools build system.
```

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,409 @@
+- ReFrame will use a user-provided Spack environment in order to build and test a set of specs.
+- Tutorial in `tutorials/build_systems/spack/spack_test.py`
+- In `rfm_job.out` you can see that it
+  - Creates a blank environment
```
Those two bullets appear at the same level as the one above. I think they need another space.

tkoskela commented:
I think I addressed your comments, thanks for those!

This is maybe a bit too much of a lecture notes format. I think I need to go through it and write some more prose

ilectra commented on docs/tutorial/durham_reframe_tutorial.md:
```diff
@@ -0,0 +1,414 @@
+
+## Configuring ReFrame for HPC systems (Cosma)
+> In ReFrame, all the details of the various interactions of a test with the system environment are handled transparently and are set up in its configuration file.
+- [Configuration](https://reframe-hpc.readthedocs.io/en/v4.5.2/tutorial_basics.html#more-of-hello-world)
```
This list appears as continuous text, probably because it's inside a comment.

nicolin commented:
Installation section:  Following on from creating the virtual env and cloning /installing the tests. It should be made clear that if the next step installing spack and reframe is done inside the virtual env or outside it.  The reason is that when you go to the next page on a new system it says  "spack env create --without-view -d /path/to/environment" .  However based on the previous instruction spack is installed in the virtual env that the excalibur-tests  were put in.   

Reframe: Running the benchmark cases, it would be expected to have it create all IO in a defined folder, providing the -o MyBenchCasesA  only puts a subset of files in there with a default/default/default folder structure. Either clarity should be given directly on this page as to the folders and structure created or point to a Reframe man page if this is a standard structure.    

This is based me as a first time user of spack and Reframe:  Use case  that I normally do  either for different systems or different configs eg GPUA GPUB on the same system is that what to have all the results in one folder so   MyBenchCasesA  MyBenchCasesB etc in a portable manner.

nicolin commented on docs/tutorial/getting-started.md:
```diff
@@ -0,0 +1,67 @@
+# Getting Started on HPC systems
+
+These tutorials have been ran in in-person workshops on various HPC systems in the UK. It should be possible to run on any of the [supported systems](../systems.md), or to [set up the tools](../setup.md) on a local machine. If you have access to one of the systems we've previously used, this tutorial helps you get set up. Otherwise, please consult the [documentation](../install.md)
```
" ran in in-person workshops", change to  "have been run at in-person workshops ...

nicolin commented on docs/tutorial/getting-started.md:
```diff
@@ -0,0 +1,67 @@
+	2. A command line terminal with an ssh client. Most Linux and Mac systems come with these preinstalled. Please see [Connecting to ARCHER2](https://docs.archer2.ac.uk/user-guide/connecting/#command-line-terminal) for more information and Windows instructions.
+
+=== "ARCHER2"
+	To run these tutorials on ARCHER2, you need to [connect to ARCHER2 via ssh](https://docs.archer2.ac.uk/user-guide/connecting/). You will need
```
will need:

nicolin commented on docs/tutorial/postprocessing_tutorial.md:
```diff
@@ -0,0 +1,204 @@
+
+### Postprocessing features 
+
+The postprocessing can be performed either on a GUI or a CLI. It takes as input either a single perflog or a path that contains perflogs, and it is driven by a configuration YAML file (more on this later). Its outputs can be csv files of the whole or filtered perflog contents, as well as plots.
```
more on this later.  put a link to the section rather

nicolin commented on docs/tutorial/setup-python.md:
```diff
@@ -0,0 +1,28 @@
+=== "Cosma"
+
+    This tutorial is run on the [Cosma](https://cosma.readthedocs.io/en/latest/) supercomputer.
+    It should be straightforward to run on a different platform, the requirements are  `gcc`, `git` and `python3`. (for the later parts you also need `make`, `autotools`, `cmake` and `spack`).
```
so any version of gcc and python3  ? Else give supported or min,max versions

tkoskela replied:
Python 3.7 or later is a requirement. For gcc or git I don't know if there's a minimum version required. I added the minimum versions supported by spack just in case.

nicolin reviewed (changes requested):
Added comments on a first glace read. Will do an actual run in the coming days.

tkoskela commented on docs/tutorial/profiling_tutorial.md:
```diff
@@ -0,0 +1,70 @@
+- Spack installs the `nvidia-nsight-systems` package in the background, including the GUI
+- The paths to the collected profile data, and to the GUI launcher are written into `rfm_job.out`
+- To run the GUI remotely, you need to login with `ssh -X`. It may be slow on a remote system.
+- You can (spack) install the GUI locally to view the data.
```
Thanks! I created #358 for this

tkoskela commented:
> Installation section: Following on from creating the virtual env and cloning /installing the tests. It should be made clear that if the next step installing spack and reframe is done inside the virtual env or outside it. The reason is that when you go to the next page on a new system it says "spack env create --without-view -d /path/to/environment" . However based on the previous instruction spack is installed in the virtual env that the excalibur-tests were put in.

I added more detail on this in the installation and setup sections of the documentation. Is it clear now?

> Reframe: Running the benchmark cases, it would be expected to have it create all IO in a defined folder, providing the -o MyBenchCasesA only puts a subset of files in there with a default/default/default folder structure. Either clarity should be given directly on this page as to the folders and structure created or point to a Reframe man page if this is a standard structure.
> 
> This is based me as a first time user of spack and Reframe: Use case that I normally do either for different systems or different configs eg GPUA GPUB on the same system is that what to have all the results in one folder so MyBenchCasesA MyBenchCasesB etc in a portable manner.

Upon reading a bit of ReFrame documentation I found the flag that does this. It is `--prefix`, not `--outputdir` :upside_down_face: I've updated this in the documentation.

nicolin commented:
When going to the tutorials section, Getting started I expected there to be a general option to the tutorials but it is a setup for two specific HPC systems. I know it says " should be straightforward to run on a different platform"  could we just make a third tab Archer, Cosma, General.   It would be mainly a copy and paste that could then have some specific options. ( PS: I am doing this on not Archer or Cosma)

tkoskela commented:
> When going to the tutorials section, Getting started I expected there to be a general option to the tutorials but it is a setup for two specific HPC systems. I know it says " should be straightforward to run on a different platform" could we just make a third tab Archer, Cosma, General. It would be mainly a copy and paste that could then have some specific options. ( PS: I am doing this on not Archer or Cosma)

While in general I support this idea, it's quite complicated (as you're probably finding out) to do a general tutorial that can be reasonably expected to work on everyone's local system. That is why we've always done the in-person tutorials on Cosma or ARCHER2 so that we provide the same environment for all attendees.

Since it's difficult to guarantee that general instructions would work, I'd rather not put them in a tutorial but keep them in the Setup for new systems documentation. This is trying to keep the [types of documentation](https://docs.divio.com/documentation-system/tutorials/) separate so that tutorials take you step by step through a task, while the Setup documentation is more of a reference guide.

In either case, I think it's too much work to be added onto this PR so I'd rather do it in a separate one.

tkoskela commented:
@nicolin see [ce213f9](https://github.com/ukri-excalibur/excalibur-tests/pull/344/commits/ce213f9c6ae777163e982642d4a02add0c8503d6)

nicolin commented:
[like]  Govender, Nicolin reacted to your message:
________________________________
From: Tuomas Koskela ***@***.***>
Sent: Monday, February 10, 2025 2:42:28 PM
To: ukri-excalibur/excalibur-tests ***@***.***>
Cc: Govender, Nicolin ***@***.***>; Mention ***@***.***>
Subject: Re: [ukri-excalibur/excalibur-tests] Update tutorials (PR #344)


⚠ Caution: External sender


@nicolin<https://github.com/nicolin> see ce213f9<https://github.com/ukri-excalibur/excalibur-tests/pull/344/commits/ce213f9c6ae777163e982642d4a02add0c8503d6>

—
Reply to this email directly, view it on GitHub<https://github.com/ukri-excalibur/excalibur-tests/pull/344#issuecomment-2648206141>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABO2PHLX7U4FYF7REQAZJZD2PC3FJAVCNFSM6AAAAABOSIZ3ZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBYGIYDMMJUGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

tkoskela commented:
Click the big green approve button? :grin:
